### PR TITLE
New Reducer GPU Indexing

### PR DESCRIPTION
**Pull request by MrBurmark**
# Use GPU Indexing info in new reduction interface.

- This PR is a feature and bugfix
- It does the following (modify list as needed):
  - Fixes direct cuda/hipMalloc call and memory leaks
  - Adds compile time block size optimization for new reducer interface at the request of myself

### CODE REVIEW

**Comment by rhornung67** [comment]
Grrfff. Something changes in the CZ GitLab instance where it thinks that any PR is a draft PR. We've seen this before and need to get to the bottom of it, or else remove that check and run CI on all PRs.

File: include/RAJA/policy/cuda/reduce.hpp [hunk 1]
```diff
@@ -641,15 +639,13 @@ RAJA_DEVICE RAJA_INLINE bool grid_reduce(RAJA::expt::detail::Reducer<OP, T>& red
...
-    temp = block_reduce<OP>(temp, OP::identity());
+    temp = block_reduce<ThreadIterationGetter, OP>(temp, OP::identity());
 
     // one thread returns value
     if (threadId == 0) {
       *(red.devicetarget) = temp;
     }
   }
-
-  return lastBlock && threadId == 0;
```
Comment (rchen20) [inline include/RAJA/policy/cuda/reduce.hpp:652]: Shouldn't we keep this? We should still want to know whether we've hit the last block?

**Review by rchen20 — APPROVED** [review]
Just one question about the return statement of `grid_reduce`, but this looks good otherwise.

File: include/RAJA/policy/cuda/reduce.hpp (hunk 1 above)
Comment (MrBurmark) [reply include/RAJA/policy/cuda/reduce.hpp:652]: That is a holdover from the non-expt version where it is used, but here it is not used. I'll clean this up in a subsequent PR where the two versions will be recombined.

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**